### PR TITLE
wayland: Initial text input and IME support

### PR DESCRIPTION
**Pull request by flibitijibibo**
Note that this is purely to make it possible to enter text that requires
composition - for example, before this commit Kanji input didn't work at all.
~XWayland does not appear to support our X implementation~, so this is critical
to have implemented in some form. EDIT: #4248 fixes XWayland.

The big problem this still has is that we need the window position, and this is
still not implemented. Once we have this information we can do the equivalent
of XTranslateCoordinates to put the rectangle where we want it.

### CODE REVIEW

**Comment by Cacodemon345** [comment]
Does this work with Squeekboard?

**Comment by sulix** [comment]
FWIW, this is a slight regression for me as-is. I'm running under KDE Plasma 5.21.3 wayland with both simple compose key use and Japanese input with fcitx/mozc.

SDL's testime works under XWayland both with and without the patch (both for the compose key and full Kanji input).

However, with SDL_VIDEODRIVER=wayland, there are some annoying side effects which may or may not be related to the candidate window positioning.

Without this patch:
- no IME input (neither compose key nor mozc input) work

With this patch
- Compose key input still doesn't work (keys are passed through as plain ASCII input)
- Attempting to type anything with mozc enabled causes a zenity dialog box to appear repeatedly. This dialog has a caption of "Last Mouse Position", and contents something like "(608, 315), rel (0, 0)".
- These dialogs reappear whenever closed, even if the testime app is minimised, and it's really difficult to close testime without kill -9 at this point.

Regardless, I'm looking forward to IME working under SDL wayland (even if it actually works a treat under XWayland for me).

**Comment by flibitijibibo** [comment]
>Attempting to type anything with mozc enabled causes a zenity dialog box to appear repeatedly. This dialog has a caption of "Last Mouse Position", and contents something like "(608, 315), rel (0, 0)".

Can you get a stack trace of this? That sounds more like the messagebox implementation we introduced recently.

For the compose key I have no idea where to look...  and for my XWayland setup it's likely that the issue is only with ibus.

**Comment by flibitijibibo** [comment]
Figured out what the XWayland deal was and opened #4248. Once that's in I'll rebase this and try to figure out how the heck compose keys work in Wayland...

**Comment by sulix** [comment]
Okay: it turned out that something (possibly three things) actually broke with my OpenSUSE/KDE setup, and IME support was broken more generally when switching between X11 and Wayland sessions.
With those issues fixed, this actually seems to be working pretty perfectly, both under XWayland and Wayland.

It's definitely noticeable that the canidate list isn't brought up in the right position, but that's a problem hardly unique to SDL under Wayland.

The messagebox issue seems, as you said, to be related to the messagebox change — I'd forgotten pressing 'Space' in the test apps brought that dialog up, and my IME trigger shortcut is Ctrl+Space. When the IME stuff was broken, the space was no longer getting eaten by fcitx, so it triggered that. There were still a couple of issues with the zenity-based messageboxes: SDL doesn't seem to get the KEYUP for space when it's no-longer focused, so keeps showing more dialogs, but that's a different issue.

tl;dr: This is actually working, sorry for the confusion!

**Comment by flibitijibibo** [comment]
Patch has been rebased, this should be ready to go now.
